javagen/src/main/java/com/azure/autorest/template/ClientMethodTemplate.java

```diff
@@ -708,6 +799,24 @@ protected void generatePagedAsyncSinglePage(ClientMethod clientMethod, JavaType
         });
     }
 
+    private static boolean responseTypeHasDeserializedHeaders(IType type) {
+        if (type instanceof GenericType && "Mono".equals(((GenericType) type).getName())) {
+            type = ((GenericType) type).getTypeArguments()[0];
+        }
+
+        // TODO (alzimmer): ClassTypes should maintain reference to any super class or interface they extend/implement.
+        // This code is based on the previous implementation that assume if the T type for Mono<T> is a class that
+        // it has deserialized headers. This won't always be the case, but ClassType also isn't able to maintain
+        // whether the class is an extension of ResponseBase.
+        if (type instanceof ClassType) {
+            return true;
+        } else if (type instanceof GenericType && "ResponseBase".equals(((GenericType) type).getName())) {
+            return true;
+        } else {
+            return false;
+        }
```

Comment:
Fluent does this.
https://github.com/Azure/autorest.java/blob/main/fluentgen/src/main/java/com/azure/autorest/fluent/util/FluentUtils.java#L320-L337

Reply:
I'm thinking something similar, I think `ClassType` may want to introduce a new field that maintains which types it's implementing or extending. This will give us a richer experience throughout the code base instead of having to potentially do another Swagger lookup or wrangling external templates.